### PR TITLE
improve user action log

### PR DESCRIPTION
**Pull request by Ewoutvans**

### CODE REVIEW

File: app/controllers/Application.scala [hunk 1]
```diff
@@ -398,16 +398,42 @@ final class Application @Inject()(data: DataHelper,
...
   }
 
-  def showLog() = showLogWithPage(0)
+  def showLog(page: Option[Int], userFilter: Option[Int], projectFilter: Option[Int], contextFilter: Option[Int]) = (Authenticated andThen PermissionAction[AuthRequest](ViewLogs)).async { implicit request =>
+    val pageSize = 50
+    val offset = page.getOrElse(0) * pageSize
+
+    val default = LiteralColumn(1) === LiteralColumn(1)
+
+    val logQuery = queryLog filter { case (action, user, project, version, page) =>
```
Comment (Katrix) [inline app/controllers/Application.scala]: Avoid infix here and a bit further down

File: app/controllers/Application.scala [hunk 2]
```diff
@@ -398,16 +398,42 @@ final class Application @Inject()(data: DataHelper,
...
+    } sortBy { case (action, user, project, version, page) =>
+      action.id.desc
+    } drop offset take pageSize
 
-  def showLogWithPage(page: Int) = (Authenticated andThen PermissionAction[AuthRequest](ViewLogs)).async { implicit request =>
-    val limit = 50
-    val offset = page * limit
     for {
-      actions <- service.access[LoggedActionModel](classOf[LoggedActionModel]).filter(u => true, limit, offset)
+      actions <- service.DB.db.run(logQuery.result)
```
Comment (Katrix) [inline app/controllers/Application.scala]: All of these can from what I can see be run in parallel. use parMapN

File: app/controllers/Application.scala [hunk 3]
```diff
@@ -398,16 +398,42 @@ final class Application @Inject()(data: DataHelper,
...
+
+  private def queryLog = {
+    val tableLoggedAction = TableQuery[LoggedActionTable]
+    val userTable = TableQuery[UserTable]
+    val projectTable = TableQuery[ProjectTableMain]
+    val versionTable = TableQuery[VersionTable]
+    val pageTable = TableQuery[PageTable]
+
+    for {
+      ((((action, user), project), version), page) <- (((tableLoggedAction joinLeft userTable on (_.userId === _.id)) joinLeft projectTable on (_._1.actionContextId === _.id)) joinLeft versionTable on (_._1._1.actionContextId === _.id)) joinLeft pageTable on (_._1._1._1.actionContextId === _.id)
```
Comment (Katrix) [inline app/controllers/Application.scala]: Avoid infix

**Review by Katrix — CHANGES_REQUESTED** [review]

File: app/controllers/Application.scala [hunk 4]
```diff
@@ -398,16 +398,43 @@ final class Application @Inject()(data: DataHelper,
...
   }
 
-  def showLog() = showLogWithPage(0)
+  def showLog(page: Option[Int], userFilter: Option[Int], projectFilter: Option[Int], contextFilter: Option[Int]) = (Authenticated andThen PermissionAction[AuthRequest](ViewLogs)).async { implicit request =>
+    val pageSize = 50
+    val offset = page.getOrElse(0) * pageSize
+
+    val default = LiteralColumn(1) === LiteralColumn(1)
+
+    val logQuery = queryLog.filter({ case (action, user, project, version, page) =>
```
Comment (Katrix) [inline app/controllers/Application.scala]: Don't use both parenthesis and braces here. Just remove the parenthesis.

File: app/controllers/Application.scala [hunk 5]
```diff
@@ -398,16 +398,43 @@ final class Application @Inject()(data: DataHelper,
...
+      action.userId.===(userFilter).getOrElse(default) &&
```
Comment (Katrix) [inline app/controllers/Application.scala]: Symbolic infix notation is fine

File: app/controllers/Application.scala [hunk 6]
```diff
@@ -398,16 +398,43 @@ final class Application @Inject()(data: DataHelper,
...
+    }).sortBy({ case (action, user, project, version, page) =>
+      action.id.desc
+    }).drop(offset).take(pageSize)
+
+    (
+      service.DB.db.run(logQuery.result),
+      service.access[LoggedActionModel](classOf[LoggedActionModel]).size,
+      request.currentUser.get can ViewIp in GlobalScope
+    ).parMapN {
+      case (actions, size, canViewIP) =>
```
Comment (Katrix) [inline app/controllers/Application.scala]: Don't think you need case here

**Review by Katrix — CHANGES_REQUESTED** [review]

**Review by Katrix — APPROVED** [review]

**Review by phase — APPROVED** [review]
:pray:

**Pull request merged**